File: QuizAppClient/src/client/ClientMain.java
package client;

import common.Question;;
import ui.ClientGui;
import ui.SearchingThread;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.net.UnknownHostException;
import java.util.*;

public class ClientMain {
    private static final String IP = "192.168.56.1";
    private static final int PORT = 5000;
    private static ArrayList<Question> questionList = new ArrayList<>();
    private static Socket socket;
    private static Integer SIZE;
    private static ObjectOutputStream outputStream;
    private static ObjectInputStream inputStream;
    private static ClientGui clientGui;
    private static String myName;
    public static String opponentName;

    public static void main(String[] args) {
        int time = 0;
        try {
            socket = new Socket(IP, PORT);
            outputStream = new ObjectOutputStream(socket.getOutputStream());
            inputStream = new ObjectInputStream(socket.getInputStream());
            socket.setSoTimeout(15000);
            clientGui = new ClientGui();
            clientGui.showNameDialog();
            while ((clientGui.isNameEntered())==false);
            myName = clientGui.getNameEntered();
            //System.out.println(myName +"Entered");
            sendNameAndGetOpponent();
            while (SearchingThread.isStartPressed()==false);

            getQuestions();
            time = getTime();
            clientGui.setQuestionList(questionList);
            clientGui.setTIME(time);
            clientGui.start();
        } catch (UnknownHostException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }
        //new ClientGui(questionList, time);
    }

    private static int getTime() throws IOException, ClassNotFoundException {
        int time = 0;
        time = (Integer) inputStream.readObject();
        return time;
    }

    private static void getQuestions() throws IOException, ClassNotFoundException {
        System.out.println("Starting to recv questions ");
        Question question;
        SIZE = (Integer) inputStream.readObject();
        for (int i = 0; i < SIZE; i++) {
            question = (Question) inputStream.readObject();
            questionList.add(question);
            System.out.println(question.getId());
        }
        System.out.println("Questions recevied");
    }
    public static int[] sendResult() {
        try {
            int score[] = new int[3];
            HashMap<Integer, String> answerList = ClientGui.getQuestioAnswer();
            //System.out.println(answerList.size());
            outputStream.writeObject(answerList);
            score[0] = (Integer) inputStream.readObject();
            score[1] = (Integer) inputStream.readObject();
            score[2] = ((Boolean) inputStream.readObject())? 1: 0;
            System.out.println("Score ="+score[0]+" "+score[1]+" "+score[2]);
            return score;
        } catch (IOException e) {
            e.printStackTrace();
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }
        return new int[]{0};
    }

    public static void sendNameAndGetOpponent(){
        try {
            outputStream.writeObject(myName);
            SearchingThread searchingThread = new SearchingThread(clientGui);
            searchingThread.start();
            opponentName = (String) inputStream.readObject();
            System.out.println("Oppoenent = "+opponentName);
            searchingThread.setOpponentFound(true);
            searchingThread.join();

            return;
        } catch (IOException e) {
            e.printStackTrace();
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return;
    }
}

File: QuizAppClient/src/client/ClientMain.kt
package client

import common.Question
import ui.ClientGui
import ui.SearchingThread
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.net.Socket
import java.net.UnknownHostException
import java.util.*

object ClientMain {
    private const val IP = "192.168.56.1"
    private const val PORT = 5000
    private val questionList = ArrayList<Question>()
    private var socket: Socket? = null
    private var SIZE: Int? = null
    private var outputStream: ObjectOutputStream? = null
    private var inputStream: ObjectInputStream? = null
    private var clientGui: ClientGui? = null
    private var myName: String? = null
    @JvmField
    var opponentName: String? = null

    @JvmStatic
    fun main(args: Array<String>) {
        var time = 0
        try {
            socket = Socket(IP, PORT)
            outputStream = ObjectOutputStream(socket!!.getOutputStream())
            inputStream = ObjectInputStream(socket!!.getInputStream())
            socket!!.soTimeout = 15000
            clientGui = ClientGui()
            clientGui!!.showNameDialog()
            while (clientGui!!.isNameEntered() == false);
            myName = clientGui!!.getNameEntered()
            //System.out.println(myName +"Entered");
            sendNameAndGetOpponent()
            while (SearchingThread.isStartPressed == false);
            questions
            time = time
            clientGui!!.questionList = questionList
            clientGui!!.tIME =  time
            clientGui!!.start()
        } catch (e: UnknownHostException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }
        //new ClientGui(questionList, time);
    }

    @get:Throws(IOException::class, ClassNotFoundException::class)
    private val time: Int
        private get() {
            var time = 0
            time = inputStream!!.readObject() as Int
            return time
        }

    @get:Throws(IOException::class, ClassNotFoundException::class)
    private val questions: Unit
        private get() {
            println("Starting to recv questions ")
            var question: Question
            SIZE = inputStream!!.readObject() as Int
            for (i in 0 until SIZE!!) {
                question = inputStream!!.readObject() as Question
                questionList.add(question)
                println(question.id)
            }
            println("Questions recevied")
        }

    @JvmStatic
    fun sendResult(): IntArray {
        try {
            val score = IntArray(3)
            val answerList: HashMap<Any?, Any?> = ClientGui.getQuestionAnswer()
            //System.out.println(answerList.size());
            outputStream!!.writeObject(answerList)
            score[0] = inputStream!!.readObject() as Int
            score[1] = inputStream!!.readObject() as Int
            score[2] = if (inputStream!!.readObject() as Boolean) 1 else 0
            println("Score =" + score[0] + " " + score[1] + " " + score[2])
            return score
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }
        return intArrayOf(0)
    }

    fun sendNameAndGetOpponent() {
        try {
            outputStream!!.writeObject(myName)
            val searchingThread = clientGui?.let { SearchingThread(it) }
            searchingThread!!.start()
            opponentName = inputStream!!.readObject() as String
            println("Oppoenent = $opponentName")
            searchingThread!!.isOpponentFound = true
            searchingThread?.join()
            return
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
        return
    }
}